### PR TITLE
fix passing of package paths+URLs to package module when using apt

### PR DESCRIPTION
**Pull request by afeld**
##### SUMMARY
While most of the packaging modules support a package file path or URL being passed to their `name`, [the `apt` module requires a different one (`deb`)](https://docs.ansible.com/ansible/latest/apt_module.html#options). This change looks for a suffix of `.deb` and adjusts the arguments accordingly. Also added some tests around argument handling in the package module.

This change was inspired by seeing [a role that was installing packages cross-platform](https://github.com/singleplatform-eng/ansible-role-nessus-agent/blob/cb21622189efabef4c17245a9addac01654c8ae8/tasks/main.yml#L3-L9). The simplified version of their code:

```yaml
- name: Install from repo
  package: name={{pkg}}
  when: "'.deb' not in pkg"

- name: Install from .deb package
  apt: deb={{pkg}}
  when: "'.deb' in pkg"
```

Following this change, it could simply be

```yaml
- name: Install from repo
  package: name={{pkg}}
```

This will make cross-platform support in Ansible code just that little bit easier.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
package

##### ANSIBLE VERSION

```
ansible 2.4.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/ubuntu/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/ubuntu/.local/lib/python2.7/site-packages/ansible
  executable location = /home/ubuntu/.local/bin/ansible
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

With the following command, run on Ubuntu 16.04:

```sh
ansible all -b -i "localhost," -c local -m package -a "name=http://mirrors.kernel.org/ubuntu/pool/universe/c/cloc/cloc_1.60-1.1_all.deb"
```

Before:

```
localhost | FAILED! => {
    "changed": false,
    "msg": "No package matching 'http://mirrors.kernel.org/ubuntu/pool/universe/c/cloc/cloc_1.60-1.1_all.deb' is available"
}
```

After:

```
localhost | SUCCESS => {
    "changed": true,
    ...
}
```

### CODE REVIEW

**Comment by afeld** [comment]
Hmm, the FreeBSD tests seem flaky...seeing other recent pull requests with the same failures. Let me know what I should do.

Re: the change, another way to approach this problem is to deprecate the `deb` option of the `apt` module, and have the `name` to handle a filename or URL. I figured that would be more controversial, so didn't do so. Let me know what you think - happy to take it as a follow-up.

**Comment by mattclay** [comment]
The FreeBSD tests have been fixed. I've restarted CI.

**Comment by mattclay** [comment]
CI failure unrelated to PR. Restarting CI.

File: lib/ansible/modules/packaging/os/package.py [hunk 1]
```diff
@@ -28,6 +28,7 @@
   name:
     description:
       - "Package name, or package specifier with version, like C(name-1.0)."
+      - This can be a package name, or with most package managers, a file path to a package or a URL.
```
Comment (gundalow) [inline lib/ansible/modules/packaging/os/package.py]: May be worth adding "From Ansible 2.5 this can also be..."

**Review by gundalow — COMMENTED** [review]
Yay for tests :)

File: lib/ansible/modules/packaging/os/package.py (hunk 1 above)
Comment (afeld) [reply lib/ansible/modules/packaging/os/package.py]: Well, it's supported URLs/paths for Yum (at least) for longer...not sure how much detail we should squeeze in there.

**Review by gundalow — APPROVED** [review]

**Comment by mattclay** [comment]
CI failure in integration tests due to traceback:

```
Traceback (most recent call last):
  File "/root/ansible/lib/ansible/executor/task_executor.py", line 137, in run
    res = self._execute()
  File "/root/ansible/lib/ansible/executor/task_executor.py", line 542, in _execute
    result = self._handler.run(task_vars=variables)
  File "/root/ansible/lib/ansible/plugins/action/package.py", line 79, in run
    new_module_args = self.new_module_args()
  File "/root/ansible/lib/ansible/plugins/action/package.py", line 40, in new_module_args
    if new_args['name'].endswith('.deb'):
AttributeError: 'list' object has no attribute 'endswith'
```

**Review by bcoca — CHANGES_REQUESTED** [review]
Actually, MOST (apk, homebrew, pkgng, pkg5, macports, etc) package modules do not support a URL.

package is meant to be the lowest common denominator, as such it should not implement options that only apply to a minority of modules

**Comment by afeld** [comment]
> MOST (apk, homebrew, pkgng, pkg5, macports, etc) package modules do not support a URL

Fair enough. Suggestions about how to reflect that in the docs? Just go with "some package managers", or try and spell out which supports which?

> package is meant to be the lowest common denominator, as such it should not implement options that only apply to a minority of modules

Understood. As I suggested [above](https://github.com/ansible/ansible/pull/34290#issuecomment-354484418), I think this logic would ideally be moved down to the `apt` module, but figured this would be a less controversial change.

**Comment by bcoca** [comment]
@afeld considering the number of package managers and the frequency in which they change, i don't think a specific list would be accurate for long.

+1 for moving this to the specific modules that actually need to support this.

**Comment by bcoca** [comment]
Hi!

Thanks very much for your submission to Ansible.  It sincerely means a lot to us that you've taken time to contribute.

Unfortunately, we are going to decline this PR because as stated above `package` is meant to handle all package managers and as such we keep it with the 'lowest common denominator' of options.

However, we're absolutely always up for discussion.  
Since this is a really busy project, we don't always see comments on closed tickets, but want to encourageopen dialog.  

You can stop by the development list, and we'd be glad to talk about it - and we might even be persuaded otherwise!

   * https://groups.google.com/forum/#!forum/ansible-devel

In the future, sometimes starting a discussion on the development list prior to implenting a feature can make getting things included a little easier, but it's not always necessary.

Thank you once again for this and your interest in Ansible!

**Pull request closed without merging**